src/remoteCollection/handleCollectResponse.js

```diff
@@ -100,92 +101,79 @@ function prepareTransformArgs(generator) {
  * Takes each subject from the generator and makes a separate request
  * to the appropriate endpoint. Then enqueues each sample with handleCollectResponse
  *
- * @param  {Promise} generatorPromise - Promise that resolves to generator object
- * @returns {Promise} - which resolves to the queue length after enqueuing all samples,
- * or an error.
+ * @param  {Object} generator - generator object
+ * @returns {Number} - the queue length after enqueuing all samples
  * @throws {ValidationError} if thrown by prepareUrl (from sendRemoteRequest).
  */
-function handleCollectResponseBySubject(generatorPromise) {
-  return generatorPromise.then((g) => {
+function handleCollectResponseBySubject(generator) {
+  return Promise.map(generator.subjects, (subject) => {
     // need to clone generator because we are doing async operation with generator data
-    for (let subject of g.subjects) {
-      const _g = JSON.parse(JSON.stringify(g));
-      _g.subjects = [subject];
-      let qLength = handleCollectResponse(prepareRemoteRequest(_g));
-
-      // return qLength if this is the final subject we enqueue
-      if (subject === g.subjects[g.subjects.length - 1]) return qLength;
-    }
-  });
+    const _g = JSON.parse(JSON.stringify(generator));
```

Comment:
Is there any code convention when using underscore for naming vars?

Reply:
It's sometimes used to indicate a temporary variable when the variable name is already taken.
This line was already there, I just changed the enclosing loop.